File: ios/QRFileTransfer/QrProtocol.swift
import Foundation
import CryptoKit
import zlib

/// QrProtocol — формат содержимого QR-кодов (единый с Android-версией).
///
/// Заголовок сессии ("t":"h"):
///   {"v":1,"t":"h","sid":"<16 hex>","name":"photo.jpg","mime":"image/jpeg",
///    "size":5242880,"sha":"<sha256 hex>","total":6554,"cs":800,"nonce":"<8 hex>"}
///
/// Фрагмент данных ("t":"d"):
///   {"v":1,"t":"d","sid":"<16 hex>","i":123,"d":"<base64>","c":"0a1b2c3d"}
///
/// Каждая сессия начинается со случайного sid, а заголовок содержит случайный
/// nonce, поэтому QR-коды всегда разные — даже для одного и того же файла.
enum QrProtocol {
    static let version = 1
    static let chunkRawSize = 800
    static let maxFileSize: Int64 = 100 * 1024 * 1024

    static func newSessionId() -> String {
        String(UUID().uuidString.replacingOccurrences(of: "-", with: "").prefix(16))
    }

    static func sha256Hex(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    static func crc32Value(_ data: Data) -> UInt32 {
        guard !data.isEmpty else { return 0 }
        return data.withUnsafeBytes { buf -> UInt32 in
            guard let base = buf.bindMemory(to: UInt8.self).baseAddress else { return 0 }
            return zlib.crc32(0, base, uInt(data.count))
        }
    }

    static func crcHex(_ value: UInt32) -> String {
        String(format: "%08x", value)
    }

    static func buildHeader(
        sid: String,
        name: String,
        mime: String,
        size: Int64,
        sha: String,
        total: Int,
        chunkRaw: Int = chunkRawSize,
        nonce: String = String(UUID().uuidString.replacingOccurrences(of: "-", with: "").prefix(8))
    ) -> String {
        let object: [String: Any] = [
            "v": version, "t": "h", "sid": sid, "name": name, "mime": mime,
            "size": size, "sha": sha, "total": total, "cs": chunkRaw, "nonce": nonce
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: object),
              let string = String(data: data, encoding: .utf8) else { return "" }
        return string
    }

    static func buildData(sid: String, index: Int, raw: Data, crc: UInt32) -> String {
        let object: [String: Any] = [
            "v": version, "t": "d", "sid": sid, "i": index,
            "d": raw.base64EncodedString(), "c": crcHex(crc)
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: object),
              let string = String(data: data, encoding: .utf8) else { return "" }
        return string
    }

    static func chunkCount(for size: Int64, chunkRaw: Int = chunkRawSize) -> Int {
        Int((size + Int64(chunkRaw) - 1) / Int64(chunkRaw))
    }
}

struct ReceivedHeader {
    let sid: String
    let name: String
    let mime: String
    let size: Int64
    let sha: String
    let total: Int
    let chunkRaw: Int
}

struct ReceivedChunk {
    let sid: String
    let index: Int
    let bytes: Data
    let crc: UInt32
}

func parseHeader(_ json: String) -> ReceivedHeader? {
    guard let data = json.data(using: .utf8),
          let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          (object["v"] as? Int) == QrProtocol.version,
          (object["t"] as? String) == "h",
          let sid = object["sid"] as? String,
          let name = object["name"] as? String,
          let size = object["size"] as? Int64,
          let sha = object["sha"] as? String,
          let total = object["total"] as? Int else { return nil }
    let mime = object["mime"] as? String ?? "application/octet-stream"
    let chunkRaw = object["cs"] as? Int ?? QrProtocol.chunkRawSize
    return ReceivedHeader(
        sid: sid, name: name, mime: mime, size: size,
        sha: sha, total: total, chunkRaw: chunkRaw
    )
}

func parseChunk(_ json: String) -> ReceivedChunk? {
    guard let data = json.data(using: .utf8),
          let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          (object["v"] as? Int) == QrProtocol.version,
          (object["t"] as? String) == "d",
          let sid = object["sid"] as? String,
          let index = object["i"] as? Int,
          let b64 = object["d"] as? String,
          let bytes = Data(base64Encoded: b64) else { return nil }
    let crc = UInt32((object["c"] as? String) ?? "0", radix: 16) ?? 0
    return ReceivedChunk(sid: sid, index: index, bytes: bytes, crc: crc)
}

File: android/app/src/main/java/com/qrfiletransfer/app/QrProtocol.kt
package com.qrfiletransfer.app

import android.util.Base64
import org.json.JSONObject
import java.security.MessageDigest
import java.util.UUID
import java.util.zip.CRC32

/**
 * QrProtocol — формат содержимого QR-кодов.
 *
 * Каждый QR-код — это компактный JSON.
 *
 * Заголовок сессии ("t":"h"):
 *   {"v":1,"t":"h","sid":"<16 hex>","name":"photo.jpg","mime":"image/jpeg",
 *    "size":5242880,"sha":"<sha256 hex>","total":6554,"cs":800,"nonce":"<8 hex>"}
 *
 * Фрагмент данных ("t":"d"):
 *   {"v":1,"t":"d","sid":"<16 hex>","i":123,"d":"<base64>","c":"0a1b2c3d"}
 *     i   — номер фрагмента (1..total)
 *     d   — base64 фрагмента файла (CHUNK_RAW_SIZE байт)
 *     c   — CRC-32 фрагмента (8 hex)
 *
 * Каждая сессия начинается со случайного sid, а заголовок содержит случайный
 * nonce, поэтому QR-коды всегда разные — даже для одного и того же файла.
 */
object QrProtocol {
    const val VERSION = 1
    const val CHUNK_RAW_SIZE = 800
    const val MAX_FILE_SIZE = 99L * 1024 * 1024

    private const val TYPE_HEADER = "h"
    private const val TYPE_DATA = "d"

    fun newSessionId(): String =
        UUID.randomUUID().toString().replace("-", "").take(16)

    fun buildHeader(
        sid: String,
        name: String,
        mime: String,
        size: Long,
        sha256: String,
        totalChunks: Int,
        chunkRaw: Int = CHUNK_RAW_SIZE,
        nonce: String = UUID.randomUUID().toString().replace("-", "").take(8)
    ): String = JSONObject()
        .put("v", VERSION)
        .put("t", TYPE_HEADER)
        .put("sid", sid)
        .put("name", name)
        .put("mime", mime)
        .put("size", size)
        .put("sha", sha256)
        .put("total", totalChunks)
        .put("cs", chunkRaw)
        .put("nonce", nonce)
        .toString()

    fun buildData(sid: String, index: Int, raw: ByteArray, crc: Int): String = JSONObject()
        .put("v", VERSION)
        .put("t", TYPE_DATA)
        .put("sid", sid)
        .put("i", index)
        .put("d", encode(raw))
        .put("c", crcHex(crc))
        .toString()

    fun crc32(raw: ByteArray): Int {
        val c = CRC32()
        c.update(raw)
        return c.value.toInt()
    }

    fun crcHex(crc: Int): String = String.format("%08x", crc)

    fun sha256Hex(bytes: ByteArray): String {
        val d = MessageDigest.getInstance("SHA-256").digest(bytes)
        return d.joinToString("") { "%02x".format(it) }
    }

    fun encode(b: ByteArray): String = Base64.encodeToString(b, Base64.NO_WRAP)
    fun decode(s: String): ByteArray = Base64.decode(s, Base64.DEFAULT)

    fun chunkCount(size: Long, chunkRaw: Int = CHUNK_RAW_SIZE): Int =
        ((size + chunkRaw - 1) / chunkRaw).toInt()

    data class Header(
        val sid: String,
        val name: String,
        val mime: String,
        val size: Long,
        val sha256: String,
        val total: Int,
        val chunkRaw: Int
    )

    fun parseHeader(json: String): Header? = try {
        val j = JSONObject(json)
        if (j.optString("t") != TYPE_HEADER || j.optInt("v") != VERSION) {
            null
        } else {
            Header(
                sid = j.getString("sid"),
                name = j.getString("name"),
                mime = j.optString("mime", "application/octet-stream"),
                size = j.getLong("size"),
                sha256 = j.getString("sha"),
                total = j.getInt("total"),
                chunkRaw = j.optInt("cs", CHUNK_RAW_SIZE)
            )
        }
    } catch (e: Exception) {
        null
    }

    data class Chunk(
        val sid: String,
        val index: Int,
        val bytes: ByteArray,
        val crc: Int
    )

    fun parseChunk(json: String): Chunk? = try {
        val j = JSONObject(json)
        if (j.optString("t") != TYPE_DATA || j.optInt("v") != VERSION) {
            null
        } else {
            Chunk(
                sid = j.getString("sid"),
                index = j.getInt("i"),
                bytes = decode(j.getString("d")),
                crc = j.optString("c", "0").toInt(16)
            )
        }
    } catch (e: Exception) {
        null
    }
}